File: languages/ruby/src/provider/TableReader.rb
require "./src/model/UserInfo.rb"
require './src/model/exception/InvalidParameterException.rb'
require './src/model/exception/DataReaderException.rb'
require './src/model/DataReader.rb'

# Classe responsavel por ler e disponibilizar os dados em um formato desejado
# @see model.DataReader
class TableReader 
    include DataReader

    # Construtor publico da classe
    # @param fileName Nome do arquivo de dados a ser lido
    # @throws DataReaderException Lancada caso nao seja possivel ler os dados corretamente
    def initialize(fileName)
        raise InvalidParameterException.new "'fileName' é nil"  if(fileName.nil?)

        @fileName = fileName
        @userInfoList = self.deserializeFile(fileName)
    end

    # Obtem o nome do arquivo de dados a ser lido
    # @return Nome do arquivo de dados a ser lido
    def getFileName
        return @fileName
    end

    # Obtem todos os dados disponiveis
    # @return Lista contendo todos os dados disponiveis
    def readAll
        return @userInfoList
    end

    # Obtem os dados disponiveis dentro de um intervalo
    # @param startIndex Inicio do intervalo
    # @param endIndex Fim do intervalo
    # @return Lista contendo todos os dados disponiveis dentro do intervalo especificado
    def read(startIndex, endIndex)
        raise InvalidParameterException.new("'startIndex' é menor que 0") if(startIndex < 0)
        raise InvalidParameterException.new("'endIndex' é menor que 0") if(endIndex < 0)
        raise InvalidParameterException.new("'startIndex' é maior ou igual á 'endIndex'") if(startIndex >= endIndex)

        return @userInfoList[startIndex..endIndex]
    end

    # Desserializa o arquivo de dados, convertendo-o em uma lista de 'UserInfo'
    # @param fileName Nome do arquivo de dados
    # @return Lista contendo os dados desserilizados
    # @throws DataReaderException Lancada caso nao seja possivel ler os dados corretamente
    def deserializeFile(fileName)
        raise InvalidParameterException("'fileName' é nil") if(fileName.nil?)

        lines = File.open(fileName).drop(1)
        
        lines.map do |line|

            values = line.chop.split(",")

            user = values[0].delete(" ")
            password = values[1].delete(" ")
            credit = values[2].delete(" ").to_f

            UserInfo.new(user, password, credit)
        end
    end
end

File: languages/python3/src/provider/TableReader.py
from src.model.UserInfo import UserInfo
from src.model.exception.InvalidParameterException import InvalidParameterException
from src.model.exception.DataReaderException import DataReaderException
from src.model.DataReader import DataReader

# Classe responsavel por ler e disponibilizar os dados em um formato desejado
# @see model.DataReader
class TableReader(DataReader):

    # Construtor publico da classe
    # @param fileName Nome do arquivo de dados a ser lido
    # @throws DataReaderException Lancada caso nao seja possivel ler os dados corretamente
    def __init__(self, fileName):
        if(fileName == None):
            raise InvalidParameterException("'fileName' é None")

        self.__fileName = fileName
        self.__userInfoList = self.__deserializeFile(fileName)

    # Obtem o nome do arquivo de dados a ser lido
    # @return Nome do arquivo de dados a ser lido
    def getFileName(self):
        return self.__fileName

    # Obtem todos os dados disponiveis
    # @return Lista contendo todos os dados disponiveis
    def readAll(self):
        return self.__userInfoList
    
    # Obtem os dados disponiveis dentro de um intervalo
    # @param startIndex Inicio do intervalo
    # @param endIndex Fim do intervalo
    # @return Lista contendo todos os dados disponiveis dentro do intervalo especificado
    def read(self, startIndex, endIndex):
        if(startIndex < 0):
            raise InvalidParameterException("'startIndex' é menor que 0")
        
        if(endIndex < 0):
            raise InvalidParameterException("'endIndex' é menor que 0")
    
        if(startIndex >= endIndex):
            raise InvalidParameterException("'startIndex' é maior ou igual á 'endIndex'")

        return self.__userInfoList[startIndex:endIndex]

    # Desserializa o arquivo de dados, convertendo-o em uma lista de 'UserInfo'
    # @param fileName Nome do arquivo de dados
    # @return Lista contendo os dados desserilizados
    # @throws DataReaderException Lancada caso nao seja possivel ler os dados corretamente
    def __deserializeFile(self, fileName):
        if(fileName == None):
            raise InvalidParameterException("'fileName' é None")

        userInfoList = []

        f = open(fileName, "r")
        f.readline()

        while(True):
            line = f.readline().replace("\n","")

            if(line == "" or line == None):
                break

            values = line.split(",")

            user = values[0].strip()
            password = values[1].strip()
            credit = values[2].strip()

            userInfo = UserInfo(user, password, credit)
            userInfoList.append(userInfo)

        return userInfoList
        

            
        